cvat-canvas/src/typescript/interactionHandler.ts

```diff
@@ -0,0 +1,264 @@
+// Copyright (C) 2020 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+
+import * as SVG from 'svg.js';
+import consts from './consts';
+import Crosshair from './crosshair';
+import { translateToSVG } from './shared';
+import { InteractionData, InteractionResult, Geometry } from './canvasModel';
+
+export interface InteractionHandler {
+    transform(geometry: Geometry): void;
+    interact(interactData: InteractionData): void;
+    cancel(): void;
+}
+
+export class InteractionHandlerImpl implements InteractionHandler {
+    private onInteraction: (shapes: InteractionResult[] | null) => void;
+    private onStopInteraction: () => void;
+    private geometry: Geometry;
+    private canvas: SVG.Container;
+    private interactionData: InteractionData;
+    private cursorPosition: { x: number; y: number };
+    private shapesWereUpdated: boolean;
+    private interactionShapes: SVG.Shape[];
+    private currentInteractionShape: SVG.Shape | null;
+    private crosshair: Crosshair;
+
+    private prepareResult(): InteractionResult[] {
+        return this.interactionShapes.map((shape: SVG.Shape): InteractionResult => {
+            if (shape.type === 'circle') {
+                const points = [(shape as SVG.Circle).cx(), (shape as SVG.Circle).cy()];
+                return {
+                    points: points.map((coord: number): number => coord - this.geometry.offset),
+                    shapeType: 'points',
+                    button: (shape.style('stroke') as any as string) === 'red' ? 2 : 0,
+                };
+            }
+
+            const bbox = (shape.node as any as SVGRectElement).getBBox();
+            const points = [bbox.x, bbox.y, bbox.x + bbox.width, bbox.y + bbox.height];
+            return {
+                points: points.map((coord: number): number => coord - this.geometry.offset),
+                shapeType: 'rectangle',
+                button: 0,
+            };
+        });
+    }
+
+    private shouldRaiseEvent(ctrlKey: boolean): boolean {
+        const { interactionData, interactionShapes, shapesWereUpdated } = this;
+        const { minVertices, enabled } = interactionData;
+
+        const minimumVerticesAchieved = typeof (minVertices) === 'undefined'
+            || minVertices <= interactionShapes.length;
+        return enabled && !ctrlKey && minimumVerticesAchieved && shapesWereUpdated;
+    }
+
+    private addCrosshair(): void {
+        const { x, y } = this.cursorPosition;
+        this.crosshair.show(this.canvas, x, y, this.geometry.scale);
+    }
+
+    private removeCrosshair(): void {
+        this.crosshair.hide();
+    }
+
+    private interactPoints(): void {
+        const eventListener = (e: MouseEvent): void => {
+            if ((e.button === 0 || e.button === 2) && !e.altKey) {
```

Comment:
It's better to add default event preventing for right mouse button, i guess.

Reply:
Sure